File: itx-go/itx-contract/queue/message/message.go
package message

import "github.com/google/uuid"

const TypePostCreated = "post.created"

type PostCreatedMessage struct {
	Type     string    `json:"type"`
	PostID   int64     `json:"postId"`
	AuthorID uuid.UUID `json:"authorId"`
}

func NewPostCreated(postID int64, authorID uuid.UUID) PostCreatedMessage {
	return PostCreatedMessage{Type: TypePostCreated, PostID: postID, AuthorID: authorID}
}

File: itx-kt/itx-contract/src/main/kotlin/io/github/chehsunliu/itx/contract/queue/Message.kt
package io.github.chehsunliu.itx.contract.queue

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
sealed interface MessageBody {
    val type: String
}

@Serializable
@SerialName("post.created")
data class PostCreatedMessageBody(
    @kotlinx.serialization.Required override val type: String = "post.created",
    val postId: Long,
    val authorId: String,
) : MessageBody {
    companion object {
        const val TYPE: String = "post.created"

        fun of(
            postId: Long,
            authorId: java.util.UUID,
        ): PostCreatedMessageBody = PostCreatedMessageBody(postId = postId, authorId = authorId.toString())
    }
}
